first version of GITFans in Oscar

ThomasBreuer opened the pull request:
move the relevant code from the GITFans repository to the `experimental` directory of Oscar,
add the example in a test file,
leave out the outdated stuff

Since the rearrangements which I made today, the test example cannot be run because of a `std::bad_alloc` error in a call to `Polymake.polytope.equal_polyhedra`.
(Yesterday I saw this error for the first time, when I tried to run the example with trivial symmetry group, but the original example did not run into an error. Now this problem seems to be reproducible with the original test example.)

ThomasBreuer committed:
first version of GITFans in Oscar

move the relevant code from the GITFans repository
to the `experimental` directory of Oscar,
add the example in a test file,
leave out the outdated stuff

fingolfin commented:
Perhaps @benlorenz or @lkastner can have with debugging the `std::bad_alloc` error?

lkastner commented:
Yes, @ThomasBreuer wrote me yesterday and I did not get back to him yet. I will debug.

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,600 @@
+
+julia> G, emb = sub(permgens...);
+
+julia> GITFans.action_on_target(Q, G)
```
```suggestion
julia> Oscar.GITFans.action_on_target(Q, G)
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,600 @@
+> whose defining rays are given by subsets `S` of the rows of the matrix `Q`,
+> such that the matrix `S` has full rank and such that the ideal `I` is
+> monomial-free (see [`is_monomial_free`](@ref)) w.r.t. the variables `x_i`
+> for which the `i`-th row of `Q` is not contained in `S`.
```
```suggestion
Return orbit representatives of the group `G` on the set of those cones
whose defining rays are given by subsets `S` of the rows of the matrix `Q`,
such that the matrix `S` has full rank and such that the ideal `I` is
monomial-free (see [`is_monomial_free`](@ref)) w.r.t. the variables `x_i`
for which the `i`-th row of `Q` is not contained in `S`.
```
and similarly elsewhere

fingolfin commented:
I don't see a `std::bad_alloc` error,  but there are still some doctest issues that would need to be adjusted

fingolfin commented:
To clarify: for this to continue, I think @jankoboehm needs to be active, or anybody else who knows about GIT fans (suggestions welcome, perhaps @wdecker has some). But Thomas can't (and shouldn't) really push it forward on his own

ThomasBreuer commented:
From my point of view, the general "show stopper" was the `std::bad_alloc` error. As far as I understand, this problem has already been fixed in Polymake, and as soon as this fix is available in Oscar, we can merge this pull request.
I will change a few ugly hacks in the code where matrix groups in characteristic zero are involved; these hacks are not necessary anymore.
Concerning the rest of the code and its further development, I agree with Max that the experts are asked.

lkastner committed:
Polytopes: Revert these changes after next libpolymake release, see #758

lkastner committed:
GITFans: Revert these changes after next libpolymake release, see #758

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,604 @@
+            if rank(current_mat) == projected_dimension &&
+               is_monomial_free(I, setdiff(1:nr_variables, i))
+                cone = Polymake.polytope.Cone(INPUT_RAYS = current_mat)
+                cone.FACETS
+                cone.RAYS
```
just to  clarify, is this simply meant to trigger some computations whose results are ignored (for now)? Or what is the motivation?

lkastner replied:
The results are only ignored at toplevel. polymake uses these results internally for checking equality of the polyhedra. The problem is that polymake creates a giant LP otherwise. In the next polymake release this is fixed and these lines can go away. I even mentioned this in the commit message to it is easy to find out what to remove once that happens.

fingolfin replied:
Thanks for the clarification!

fingolfin replied:
Of course now I am also wondering why this code directly calls `Polymake.polytope.Cone` instead of using an `Oscar.Cone` object (answer is probably: because this code predates `Oscar.Cone` ;-) but then perhaps it should be brought forward to use that? Alternatively, perhaps a TODO comment to that effect ought to be inserted?

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,604 @@
+
+julia> G, emb = sub(permgens...);
+
+julia> GITFans.action_on_target(Q, G)
```
This gives an error now. I guess that means my change suggestion (which was applied) was bad and we need to add the `Oscar.` back...

ThomasBreuer committed:
polished the code a bit

- addressed the comment about the format of docstrings
- adjusted the examples in docstrings, such that doctests pass
- instead of the `GL(n, ZZ)` hack, create a matrix group over `QQ`

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+  like those shown in the paper
+  [J. Boehm, S. Keicher, Y. Ren:
+  Computing GIT-Fans with Symmetry and the Mori Chamber Decomposition of M06bar](https://arxiv.org/abs/1603.09241).
+  ([Math. Comp. 89 (2020),
```
Move this to the bib file?

ThomasBreuer replied:
Yes.

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+    for i in bitset
+        push!(return_list, perm[i])
+    end
+    return BitSet(return_list)
```
```suggestion
    return BitSet(perm[i] for i in bitset)
```
Or replace `perm` by a `PermGroupElem` and then just write `bitset^perm`. :-)

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+        end
+    end
+    return true
+end;
```
```suggestion
end
```
Several similar instances of redundant semicolons should also be updated

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+
+function bitlist_oper_tuple(bitset_tuple, perm_tuple)
+    return map(i -> bitlist_oper(bitset_tuple[i], perm_tuple[i]),
+               1:length(bitset_tuple))
```
```suggestion
    return [bitlist_oper(bitset_tuple[i], perm_tuple[i]) for 1:length(bitset_tuple)]
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+
+function find_smallest_orbit_element(elem, ggens, action, comparator, leq)
+    current_orbit = orbit(elem, ggens, action, comparator)
+    sorted_orbit = sort(current_orbit; lt=leq)
+    return sorted_orbit[1]
```
Perhaps this works, too?
```suggestion
    return minimum(current_orbit)
```

ThomasBreuer replied:
We want the smallest entry w.r.t. `leq`.
I did not find a way to compute `minimum` for an array w.r.t. a prescribed comparison function.

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+        orbs = GAP.Globals.Orbits(G.X, GAP.GapObj(set, recursive = true), GAP.Globals.OnSets)
+#T TODO: switch to an Oscar `orbits` function
+
+        for i in [Vector{Int64}(orbs[i][1]) for i in 1:length(orbs)]
```
This can be transformed first to
```suggestion
        for i in [Vector{Int64}(orb[1]) for orb in orbs]
```
and then to the following, which avoids creating a temporary `Vector`
```suggestion
        for orb in orbs
          i = Vector{Int64}(orb[1])
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+function hashes_to_polyhedral_fan(orbit_list, hash_list, hom)
+    # translate the descriptions of the orbit repres. of maximal cones
+    # to cone objects
+    result_cones = map(x -> hash_to_cone(orbit_list, x), hash_list)
```
```suggestion
    result_cones = [hash_to_cone(orbit_list, x) for x in hash_list]
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+                      for cone in result_cones]
+
+    # the set of rays
+    allrays = sort(collect(Set(vcat(rays_maxcones...))))
```
What is the `collect(Set(...))` bit about? Is this to get a duplicate free list? Then perhaps `unique` could be used instead? And then `sort! instead of `sort` to avoid yet another allocation?
```suggestion
    allrays = sort!(unique(vcat(rays_maxcones...)))
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+    R = Oscar.base_ring(SingI)
+    Rgens = Oscar.gens(R)
+    nr_variables = length(Rgens)
+    poly_list = map(f -> evaluate(f, vars_to_zero, fill(R(0), length(vars_to_zero))), gens(SingI))
```
```suggestion
    poly_list = [evaluate(f, vars_to_zero, fill(R(0), length(vars_to_zero))) for f in gens(SingI)]
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+    #   in the matrix of rays.
+    pm_rays = Polymake.Matrix(hcat(allrays...)')
+    matgens = [image(hom, x) for x in gens(domain(hom))]
+    mats_transp = [GAP.gap_to_julia(Matrix{Rational{BigInt}}, x.X)' for x in matgens]
```
```suggestion
    mats_transp = [Matrix{Rational{BigInt}}(x.X)' for x in matgens]
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+    rewr = [as_permutation(x, allrays, actfun, ==) for x in mats_transp]
+
+    rewr2 = [Vector(x) .- 1 for x in rewr]  # zero based
+    pm_gens = @Polymake.convert_to Array{Array{Int}} rewr2
```
Julia best practice is to write `MODULE.@MACRO` not `@MODULE.MACRO`
```suggestion
    pm_gens = Polymake.@convert_to Array{Array{Int}} rewr2
```

fingolfin commented on experimental/GITFans/GITFans.jl:
```diff
@@ -0,0 +1,601 @@
+    rewr2 = [Vector(x) .- 1 for x in rewr]  # zero based
+    pm_gens = @Polymake.convert_to Array{Array{Int}} rewr2
+
+    pm_mc_reps = @Polymake.convert_to Array{Array{Int}} index_result_cones
```
```suggestion
    pm_mc_reps = Polymake.@convert_to Array{Array{Int}} index_result_cones
```

fingolfin reviewed (commented):
Left some remarks... you may not wish to address all of them now, but perhaps we can at least add some notes to remind about them?

All in all, this code is still quite some "un-julia-ish".

ThomasBreuer commented:
@fingolfin Thanks for the comments. I will address most of them now.

ThomasBreuer committed:
added the bibl. entry [BKR20] ...

... and adjusted some other entries to what Documenter can process

The $\overline$ in the title causes problems, I do not have an idea how to fix this.

ThomasBreuer committed:
addressed the hints about better Julia syntax

(The `@cite` calls in the module header and in another docstring,
instead of mentioning the paper explicitly,
are not really helpful in the Julia session.)

fingolfin committed:
Merge branch 'master' into TB_GITFans_in_Oscar

fingolfin reviewed (approved):
Let's merge this now (assuming the tests still pass -- I just had to resolve a merge conflict caused by another PR being merged) and improve it further later. Thanks!
